src/rust/engine/graph/src/entry.rs

```diff
@@ -108,20 +110,38 @@ pub(crate) enum EntryState<N: Node> {
   // The previous_result value is _not_ a valid value for this Entry: rather, it is preserved in
   // order to compute the generation value for this Node by comparing it to the new result the next
   // time the Node runs.
-  NotStarted {
+  //
+  // Similarly, because exceptions are not cached in the graph, previous_error does not affect the
+  // value for this Entry, and is only used in .peek() so we can e.g. generate a scheduler stack
```

Comment:
Mm... this `peek` case came up in #9015. I had a separate resolution there, which was to associate each result with a `Session`, and then check the session during `peek`: https://github.com/pantsbuild/pants/pull/9015/files#diff-75bfebc7637079e53ddae93c399f8495R86-R92

I expect that merging these two patches together might involve replacing this error field with that facility. But I think that this works for now: thank you for adding a comment to explain it!

On a related note though: why is `previous_result` still a `Result` here?

Reply:
I think it can avoid being a `Result`! Considering the rest of your comment now!